Lib/zipfile.py

```diff
@@ -484,6 +485,13 @@ def _decodeExtra(self):
                 except struct.error:
                     raise BadZipFile(f"Corrupt zip64 extra field. "
                                      f"{field} not found.") from None
+            elif tp == 0x7075:
+                data = extra[4:ln+4]
+                # Unicode Path Extra Field
+                up_version, up_name_crc = unpack('<BL', data[:5])
+                up_unicode_name = data[5:].decode('utf-8')
+                if up_version == 1 and up_name_crc == self.orig_filename_crc:
+                    self.filename = up_unicode_name
```

Comment:
Some sanitation of filenames occur in `ZipInfo.__init__()`

https://github.com/python/cpython/blob/9ff84c7e331fa75a8c93a44cbed2bd0035889652/Lib/zipfile.py#L349-L356

Can you put those checks into a separate method and apply them here too?